Implemented the k-means++ cluster initialization technique, and some bug fixes

kirillvasilenko opened the pull request:
Here:
1. k-means++ cluster initialization
2. Some fixes for Silhouette Coefficient calculation
3. A fix for AbstractPeriodicProgressTracker

Please, look at:
1. The commit messages. I described there what I have done in a particular commit.
5. My comments in the PR.

Please, keep in mind that it is not ready to be merged yet. We need to agree that the implementation is ok, so I will remove some benchmarks and "hacky" params in the public methods.

I am not sure whether we should keep both random and k-means++ implementations, or just k-means++. So it is an open question.

If we would like to keep both implementations, I will make "proper" changes in the API. Current boolean param `initializeCentroidsWithKMeansPlusPlus` is just a hacky way to compare both implementation in the benchmark.

kirillvasilenko commented on vectoriadb-index/src/main/java/jetbrains/vectoriadb/index/Quantizer.java:
```diff
@@ -30,10 +30,10 @@ public interface Quantizer extends AutoCloseable {
 
     IntArrayList[] splitVectorsByPartitions(int numClusters, int iterations, DistanceFunction distanceFunction,
-                                            @NotNull ProgressTracker progressTracker);
+                                            boolean initializeCentroidsWithKMeansPlusPlus, @NotNull ProgressTracker progressTracker);
```
This `initializeCentroidsWithKMeansPlusPlus` param is temporary, just for the benchmark to assess the difference between the random and k-means++ initialisation methods.

When we decide that k-means++ is good enough, I will remove these params and leave only the k-means++ initialisation implementation.

I do not see much sense in keeping both implementations. But if you think we should keep both of them, I will refactor the API more carefully.

kirillvasilenko commented on vectoriadb-index/src/test/java/jetbrains/vectoriadb/index/L2PQKMeansTest.java:
```diff
@@ -132,6 +131,105 @@ public void testPQKMeansQuality() throws Exception {
     }
 
+    @Test
+    public void randomClusterInitializationVsKmeansPpBenchmark() throws Exception {
```
For this benchmark, I get results like
`meanSilhCoeffDiff: 0,006831, meanSilhCoeffRandom: 0,098650, meanSilhCoeffPP: 0,105481`

kirillvasilenko replied:
So it seems that k-means++ does a bit better.

kirillvasilenko commented on vectoriadb-index/src/test/java/jetbrains/vectoriadb/index/L2PQKMeansTest.java:
```diff
@@ -132,6 +131,105 @@ public void testPQKMeansQuality() throws Exception {
+    }
+
+    @Test
+    public void silhouetteVsSilhouetteFixedBenchmark() throws Exception {
```
For this benchmark, I get results like
```
silhouetteCoefficient = 0,196571
silhouetteCoefficientFixed = 0,109737
```

kirillvasilenko replied:
I am going to remove this benchmark, when we decide what silhouetteCoefficient implementation we keep.

kirillvasilenko commented:
I cleaned up the Silhouette Coefficient calculation code and added the simplified/medoid-based version. It does the job in O(vectors.count), so we can run it in tests on Sift1M.

I added this simplified implementation to the benchmark `silhouetteVsSilhouetteFixedBenchmark` that compares all the implementations we have. A typical result I get running the benchmark is 
```
silhouetteCoefficient = 0,188521
silhouetteCoefficientFixed = 0,098324
silhouetteCoefficientMedoid = 0,149448
```

laa reviewed (commented):
Oh yep. Thank you, Kiill.

laa reviewed (commented):
Cool, thanks

laa reviewed (commented):
Cool, thanks.

laa reviewed (changes requested):
Could you explain why you removed the part related to the choosing by weighted probably based on distance and chose to use only the vector with the largest distance instead?

laa commented:
I like this change overall. Let us discus the question left. Thank you for all the bug fixes.

laa commented:
> @laa, I implemented choosing a vector to become a new centroid based on the weighted probability proportional to `distanceToClosestCentroid^2`.
> 
> The results (Silhouette Coefficient) in my small benchmark (look at `randomClusterInitializationVsKmeansPpBenchmark`) became a bit worse than when we used the vector with the largest distance to already chosen centroids. The results are still better than random, though.
> 
> A typical comparison looks like this
> 
> ```
> max distance:          meanSilhCoeffDiff: 0,006786, meanSilhCoeffRandom: 0,148547, meanSilhCoeffPP: 0,155333
> weighted probability:  meanSilhCoeffDiff: 0,003135, meanSilhCoeffRandom: 0,149456, meanSilhCoeffPP: 0,152591
> ```
> 
> It is kind of expected, I guess. Because with weighted probability, we still may pick up not optimal/unfortunate vectors to become centroids.
> 
> What do you think?

Hi Kirill. 
Can we try on more datasets, using different distribution low, for example? Apache RNG allows the use of other distributions.

laa commented:
In general, if data has varying data densities choosing of "farthest centroid" can lead to choosing of centroids in sparse regions when there are no meaningful clusters, but can also lead to a bit worse results of course. I am curious wether we can rerpdoce this problem on practice with "farthest centroid" strategy choosen.

laa commented:
Could you also test on sif1m and gist data distributions? You can find examples of such benchmarks in the bench package.

kirillvasilenko commented:
Hmmm...

@laa, what do you think if:
1. I keep three existing strategies: random, k-means++ weighted randomness and k-means++ farthest vector.
2. I add a `centroidInitializer` param to the needed methods. Like we do for `distanceFunction`. So, we can use/experiment with all the initialization strategies we like.
3. I make the k-means++ weighted probability default option.
4. We merge the changes to let the AI team benefit from the improvements.
5. I will do some experiments around the initialization strategies.

This way, we will deliver the improvements to current customers faster and will be able to experiment afterwards.

laa commented:
@kirillvasilenko  That is a good idea. Let me know when the code is ready.

kirillvasilenko commented:
@laa, that is what I have done in the last three commits:

### Cluster initialization logic

1. Extracted the cluster initialization logic from L2PQQuantizer and arranged it as separate classes. Currently, we have three strategies: random, k-means++ (it is possible to set a randomness source and power to calculate weighted probability) and "max distance to the closest cluster".
2. I did not find enough motivation to add a param `clusterInitializer` to the Quantizer's methods. I just do not see when we want to use different strategies with a single quantizer instance in production. We want to use different strategies with a single quantizer instance in tests/benchmarks, though. So I added a possibility to set a ClusterInitializer instance via the L2PQQuantizer constructor (and a method to change it after the construction, just for tests).

I did a smoke test for all the strategies on Sift1M, 40 clusters, 10 times and that what I got
```
k-means++ - random: 0,002463, max distance - random: 0,011972, meanSilhCoeffRandom: 0,141224, meanSilhCoeffPP: 0,143687, meanSilhCoeffMaxDistance: 0,153196
```

It requires further investigation but most probably this dataset does not have sparse regions (as you noticed) so the max distance works better for it.

As you mentioned above, it is worth playing with randomness source, maybe the number of clusters, other datasets and other ways to assess the quality of clusterization (currently, we use Silhouette Coefficient Medoid, but there are more).

By the way, I am going to do a smoke test on GIST1M.

### Cleaning up

1. I  removed the old implementation of the Silhouette Coefficient and the benchmark I used to assess it.
2. Moved `symmetricDistance(...)` method from `L2PQQuantizer` to `AbstractQuantizer` as it is not L2P specific.

laa commented:
@kirillvasilenko I am astonished by your coding culture. Thank you.

kirillvasilenko commented:
@laa, I am glad you think so :)

I have done a smoke benchmark on GIST1M and the result is the following

```
k-means++ - random: 0,003709, max distance - random: 0,009039, meanSilhCoeffRandom: 0,090460, meanSilhCoeffPP: 0,094169, meanSilhCoeffMaxDistance: 0,099499
```

Looks very similar to what we got on SIFT1M.

Most probably, I will check the influence of randomness sources a bit later. As we have observed the same pattern using all the datasets SIFT10K, SIFT1M and GIST1M, I will check all the randomness sources on SIFT10K to make it faster. I see the process like this:
1. Find the best number of clusters for SIFT10K, SIFT1M and GIST1M using the Silhouette Coefficient (the best number will give the highest coefficient value).
2. Using this number of clusters, check all the randomness sources on SIFT10 and pick up the 3 best.
3. Try the best 3 randomness sources on SIFT1M and GIST1M.
4. Choose the best one and use it as a default randomness source for k-means++

As for now, if you think this code is ready to go to the master, we can merge it :)

laa commented on vectoriadb-index/src/main/java/jetbrains/vectoriadb/index/KMeansPlusPlusClusterInitializer.java:
```diff
@@ -0,0 +1,137 @@
+                distancesToClosestCentroid.setAtIndex(ValueLayout.JAVA_FLOAT, vectorIdx, Float.MAX_VALUE);
+            }
+
+            var cores = (int) Math.min(Runtime.getRuntime().availableProcessors(), numVectors);
```
Hi Kirill.

Let me outline the implemented algorithm, if I understood it correctly:

1. Choose centroids at random.

In cycle:
2. Choose the n-th centroid.

For all vectors
3. If the distance from the vector to the chosen centroid is the closest, remember it in an array of the closest distances.
4. Calculate the sum of the closest squire distances.

5. Choose a new vector for the n-th centroid using weighted probability over squire distances.


K-Means ++

In a loop:

1. Choose the point that is not a centroid at the moment
2. Find the squire distance to the closest centroid from all centroids.

3. Choose a new vector for the n-th centroid using weighted probability over squire distances.
4. Repeat the loop till not all centroids are chosen

I have a perception that implementation is incorrect because:

1. Check (1) described in the algorithm not implemented.
2. We do not base our choice on the closest distance to **all** centroids because the closest distance is unknown on the n-th step. We must repeat the algorithm n-th times (till the end when all new centroids are already chosen) to know the distance to the closest centroid.

Please prove to me that I am wrong.

kirillvasilenko replied:
Hmmm...

> 1. Check (1) described in the algorithm not implemented.

That is a good catch! 

Actually, we do this check implicitly. Here is what I mean:
* When we choose the first centroid randomly, it is obviously not a centroid because there are no centroids yet.
* When we choose all the other centroids, the min distance for all already chosen vectors-centroids will be 0 because we calculate the distance for them with themselves. So when they take part in the random selection, they are ignored because their distances = 0. A distance that is 0 does not affect the counter that is used to find a new centroid.

I see the only option when we can choose a single vector twice to be a centroid. We need the following conditions:
* Random generator gives us the biggest possible value that is = the whole sum
* The last vector in the list is already chosen as a centroid
* The error in the floating-point sum operations gives us a sum that is less than the real sum, so we have to take the list vector blindly.

As we can guess, it is hardly possible. But I agree, it is still possible, so we have to explicitly make sure that will not happen.

I will fix it.

> 2. We do not base our choice on the closest distance to all centroids because the closest distance is unknown on the n-th step. We must repeat the algorithm n-th times (till the end when all new centroids are already chosen) to know the distance to the closest centroid.

We do base our choice on the closest distance to already chosen centroids.

For this purpose, we use `distancesToClosestCentroid` array. Initially, all the distances `infinity` because we do not have any centroids. But on every step, we calculate the distance from every vector to the centroid chosen in the previous step and update `distancesToClosestCentroid` array if it is necessary (if the distance becomes smaller).

So here, it seems ok to me.

kirillvasilenko replied:
Checking that we do not make a single vector centroid twice is done.

laa replied:
Got it. You are right about distance.

laa reviewed (changes requested):
Hi, Kirill.
Please check the comment in the code and prove I am wrong.

laa reviewed (changes requested):
Hi, Kirill.
Only a small observation left.

kirillvasilenko commented on vectoriadb-index/src/main/java/jetbrains/vectoriadb/index/KMeansPlusPlusClusterInitializer.java:
```diff
@@ -131,6 +133,12 @@ public void initializeCentroids(
                             vectorToBecomeCentroidIdx++;
                         }
+                        // make sure we do not make a single vector centroid twice
+                        while (vectorsBecameCentroids.contains(vectorToBecomeCentroidIdx)) {
```
Ohhh... Yeah, I see, thanks! Done.

laa commented on vectoriadb-index/src/main/java/jetbrains/vectoriadb/index/KMeansPlusPlusClusterInitializer.java:
```diff
@@ -0,0 +1,144 @@
+                        var randomSum = rng.nextDouble(sumSquaredDistanceToClosestCentroid);
+                        double runningSum = 0;
+                        long vectorToBecomeCentroidIdx = 0;
+                        while (vectorToBecomeCentroidIdx < numVectors - 1) {
```
Hi, Kiriill.
Code looks good now.
But still performance implications left.
This cycle have O(N*k) complexity and too heave for single thread processing.
Please try to apply map-reduce approach on calculation of result of distribution based on weighted probably. Otherwise users will wait too long for this loop to be finished for all centroids.

laa replied:
Hi Krill,
Please do not work on this until you finish your current task.
It is much more important.

kirillvasilenko replied:
I got it. Thanks for noticing! I will fix it after the dotProduct quantizer task.

laa reviewed (changes requested):
Hi Kiriil.
I forogot while time passed that it is major challenge that I am expecting you to solve.
Please check comment in the code.

kirillvasilenko commented:
@laa, what do you think if we:
1. Make the default centroid initializer Random - to avoid the performance penalty until I make the weighted probability calculation scalable based on the number of CPUs.
2. Merge this branch to the master and close the merge request - to make it easier to merge. The sooner to merge, the easier.
3. I will fix the scalability issue and make the default centroid initializer K-means++ in a separate branch later (when I finish the dotProduct quantizer task).

laa commented:
@kirillvasilenko That is a wonderful idea, thank you.

kirillvasilenko commented:
@laa, I made RandomClusterInitializer the default one for L2PQQuantizer. So, it is ready to be merged.

Do I understand correctly that I should not merge it myself, but you will merge it to the master yourself?

laa commented:
@kirillvasilenko exactly. You are working on the most important issue as of now. It would be best if you were not distracted

laa commented:
I am more than a pro for merging.  If you have time, I would appreciate it.

kirillvasilenko commented:
Transparency message.

The changes from this branch were merged into the [dot-product-quantizer](https://github.com/JetBrains/xodus/tree/dot-product-quantizer) branch. So the changes will be merged in the scope of that branch.
